lib/evmone/eof.cpp

```diff
@@ -0,0 +1,160 @@
+                if (section_headers[CODE_SECTION] != 0)
+                    return {{}, EOFValidationError::multiple_code_sections};
+                state = State::section_size;
+                break;
+            default:
+                return {{}, EOFValidationError::unknown_section_id};
+            }
+            break;
+        }
+        case State::section_size:
+        {
+            const auto size_hi = *it++;
+            if (it == container_end)
+                return {{}, EOFValidationError::incomplete_section_size};
+            const auto size_lo = *it++;
+            const auto section_size = static_cast<uint16_t>((size_hi << 8) | size_lo);
+            if (section_size == 0)
+                return {{}, EOFValidationError::zero_section_size};
+
+            section_headers[section_id] = section_size;
+            state = State::section_id;
+            break;
+        }
+        case State::terminated:
+            return {{}, EOFValidationError::impossible};
+        }
+    }
+
+    if (state != State::terminated)
+        return {{}, EOFValidationError::section_headers_not_terminated};
+
+    const auto section_bodies_size = section_headers[CODE_SECTION] + section_headers[DATA_SECTION];
+    const auto remaining_container_size = container_end - it;
+    if (section_bodies_size != remaining_container_size)
+        return {{}, EOFValidationError::invalid_section_bodies_size};
+
+    return {section_headers, EOFValidationError::success};
+}
+
+std::pair<EOF1Header, EOFValidationError> validate_eof1(bytes_view container) noexcept
+{
+    const auto [section_headers, error] = validate_eof_headers(container);
+    if (error != EOFValidationError::success)
+        return {{}, error};
+
+    const EOF1Header header{section_headers[CODE_SECTION], section_headers[DATA_SECTION]};
+    return {header, EOFValidationError::success};
+}
+}  // namespace
+
+size_t EOF1Header::code_begin() const noexcept
+{
+    assert(code_size != 0);
+
+    if (data_size == 0)
+        return 7;  // EF + MAGIC + VERSION + SECTION_ID + SIZE + TERMINATOR
+    else
+        return 10;  // EF + MAGIC + VERSION + SECTION_ID + SIZE + SECTION_ID + SIZE + TERMINATOR
+}
+
+bool is_eof_code(bytes_view code) noexcept
+{
+    return code.size() > 1 && code[0] == FORMAT && code[1] == MAGIC;
+}
+
+EOF1Header read_valid_eof1_header(bytes_view::const_iterator code) noexcept
+{
+    EOF1Header header;
+    const auto code_size_offset = 4;  // FORMAT + MAGIC + VERSION + CODE_SECTION_ID
+    header.code_size =
+        static_cast<uint16_t>((code[code_size_offset] << 8) | code[code_size_offset + 1]);
+    if (code[code_size_offset + 2] == 2)  // is data section present
+    {
+        const auto data_size_offset = code_size_offset + 3;
+        header.data_size =
+            static_cast<uint16_t>((code[data_size_offset] << 8) | code[data_size_offset + 1]);
+    }
+    return header;
+}
+
+uint8_t get_eof_version(bytes_view cont) noexcept
```

Comment:
```suggestion
uint8_t get_eof_version(bytes_view container) noexcept
```

Reply:
Renamed. (I originally wanted to avoid line wrapping.)